docs/foreign_code.md

```diff
@@ -0,0 +1,156 @@
+Any time that a user has to hand-write code at a bindings layer to get memory safety right, we can assume that they will get it wrong and so we aim to automate as much as possible.
+
+The desire for sandboxing also leads to the conclusion that a 'foreign function interface' is exposing the wrong level of abstraction.
+What sandbox should a function run in?
+Our prior [CHERI-JNI](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf) work showed that treating libraries (in the loose sense, including any dependent libraries) as the unit of foreign code provided an abstraction that fitted well with sandboxing and allowed the user to determine the degree of state sharing between foreign code invocations.
+This abstraction also has software engineering benefits.
+LLVM, for example, does not have stable ABI (or API) guarantees across programs and so can cause linking errors if, for example, an application depends on one library that uses libclang from LLVM 10 to parse syntax highlighting and another that uses the JIT from LLVM 11 for GPU shaders.
+In a Verona world, these would be in different sandboxes with different 'global' namespaces and so would not conflict.
+
+### Sandboxes are surfaced as regions
+
+Sandboxing is a security *policy* that can be implemented with a variety of *mechanisms*.
+Sandboxing runs a component with a reduced set of privileges and limits its interactions with the outside world.
+It can be enforced with a separate process that runs with reduced OS privilege (e.g. the OpenSSH privesp model or Chromium's renderer sandboxes), with SFI techniques such as those used in WebAssembly compilers or NaCl (potentially with hardware acceleration from features such as Intel MPK), or with in-address-space hardware isolation such as CHERI.
+
+Sandboxes are difficult to program with in C-like languages because there is no useful construct in the abstract machine that can represent them.
+Web browsers are the most aggressively compartmentalised applications and yet run with a handful of different sandbox types.
+Because there is no abstraction in C for distinct sets of objects with mediated entry points, sandboxing code written in C leaks mechanism into the security code (for example, the idea that sandboxes are process).
+Any call between sandboxes must go via some RPC layer.
+Allocating memory inside a sandbox from outside and directly manipulating data structures is difficult to get right because the language doesn't impose any restrictions on where any given pointer can be stored.
+
+A Verona region is a set of related objects, with arbitrary topology, that have a single entry point.
+Sandboxes map cleanly to this abstraction.
+A sandbox's heap can contain pointers to any other data structures owned by the sandbox.
+It cannot contain pointers outside of the sandbox because following them would violate the sandbox policy.
+The Verona type system enforces this automatically by disallowing cross-region pointers.
```

Comment:
This is very implementation-specific.  With CHERI, you need to carve out a range of virtual address space.  With the process-based prototype, that's currently the case but it would be relatively easy to extend so that the heap is contiguous in the sandbox side and non-contiguous on the parent side, because the Verona compiler would already need some special logic for handling pointers within the sandbox (validating that they were in the range, at least), so could also handle a small set of non-contiguous mappings.

It's not clear to me that that this is actually useful though, because in practice it's typically important to impose a maximum memory limit on sandboxes to avoid DoS attacks.  In WebAssembly, this is intrinsic to the abstract machine (wasm memory regions are fixed size and the C/C++ lowering to WebAssembly places the entire heap in a single wasm region)

Reply:
This is true for web browsers that keep the memory while the page is open. But what about system software that allocates a bunch of stuff, use it and then throw it away. We need an elastic approach to memory allocation by giving back to the OS the memory that is no longer in use, even if just temporary.